c10/util/intrusive_ptr.h

```diff
@@ -280,17 +277,17 @@ class intrusive_ptr final {
   void reset_() noexcept {
     if (target_ != NullType::singleton() &&
         detail::atomic_refcount_decrement(target_->refcount_) == 0) {
-      // justification for const_cast: release_resources is basically a
-      // destructor and a destructor always mutates the object, even for const
-      // objects. NOLINTNEXTLINE(clang-analyzer-cplusplus.NewDelete)
-      const_cast<std::remove_const_t<TTarget>*>(target_)->release_resources();
-
       // See comment above about weakcount. As long as refcount>0,
       // weakcount is one larger than the actual number of weak references.
       // So we need to decrement it here.
       if (target_->weakcount_.load(std::memory_order_acquire) == 1 ||
           detail::atomic_weakcount_decrement(target_->weakcount_) == 0) {
         delete target_;
+      } else {
+        // justification for const_cast: release_resources is basically a
+        // destructor and a destructor always mutates the object, even for const
+        // objects. NOLINTNEXTLINE(clang-analyzer-cplusplus.NewDelete)
+        const_cast<std::remove_const_t<TTarget>*>(target_)->release_resources();
```

Comment:
This code movement is not sound. We MUST NOT run the weak count decrement before we `release_resources`, because once we reduce the weak count, it is possible for `release_resources` to reentrantly trigger the deallocation of the object we are working on. The order of operations looks like this:

1. On entry to reset_(), refcount == 1, weakcount == 2
2. On call to release_resources(), refcount == 0, weakcount == 1
3. Inside release_resource(), we free to weak reference to this object (in the failing test case, this happens when we deallocate autograd meta). The logic for weak reference deallocation says if weakcount == 1 I can deallocate. So deallocation happens immediately
4. We continue running release_resource() but `*this` has been deallocated out from under us and thus we ASAN fail

Reply:
So I think what we do is we swing the weakcount decref after this line, inserting here:

```
if (detail::atomic_weakcount_decrement(target_->weakcount_) == 0) {
  delete target_;
}
```

This is a pessimization if another thread races weak ref destruction but I doubt this ever happens.

Reply:
let's add a test for this lol